pyasdf/tags/transform/polynomial.py

```diff
@@ -0,0 +1,121 @@
+
+import numpy as np
+from numpy.testing import assert_array_equal
+
+from astropy import modeling
+
+from ... import yamlutil
+
+from .basic import TransformType
+
+
+__all__ = ['ShiftType', 'ScaleType', 'PolynomialType']
+
+
+class ShiftType(TransformType):
+    name = "transform/shift"
+    types = [modeling.models.Shift]
+
+    @classmethod
+    def from_tree_transform(cls, node, ctx):
+        offset = node['offset']
+        if not np.isscalar(offset):
+            raise NotImplementedError(
+                "Pyasdf currently only supports scalar inputs to Shift transform.")
+
+        return modeling.models.Shift(offset)
+
+    @classmethod
+    def to_tree_transform(cls, model, ctx):
+        return {'offset': model.offset.value}
+        #return yamlutil.custom_tree_to_tagged_tree(node, ctx)
+
+    @classmethod
+    def assert_equal(cls, a, b):
+        # TODO: If models become comparable themselves, remove this.
+        assert (isinstance(a, modeling.models.Shift) and
+                isinstance(b, modeling.models.Shift))
+        assert_array_equal(a.offset.value, b.offset.value)
+
+
+class ScaleType(TransformType):
+    name = "transform/scale"
+    types = [modeling.models.Scale]
+
+    @classmethod
+    def from_tree_transform(cls, node, ctx):
+        factor = node['factor']
+        if not np.isscalar(factor):
+            raise NotImplementedError(
+                "Pyasdf currently only supports scalar inputs to Scale transform.")
+
+        return modeling.models.Scale(factor)
+
+    @classmethod
+    def to_tree_transform(cls, model, ctx):
+        node = {'factor': model.factor.value}
+        return yamlutil.custom_tree_to_tagged_tree(node, ctx)
+
+    @classmethod
+    def assert_equal(cls, a, b):
+        # TODO: If models become comparable themselves, remove this.
+        assert (isinstance(a, modeling.models.Scale) and
+                isinstance(b, modeling.models.Scale))
+        assert_array_equal(a.factor, b.factor)
+
+
+class PolynomialType(TransformType):
+    name = "transform/polynomial"
+    types = [modeling.models.Polynomial1D, modeling.models.Polynomial2D]
+
+    @classmethod
+    def from_tree_transform(cls, node, ctx):
+        coefficients = node['coefficients']
+        n_dim = coefficients.ndim
+
+        if n_dim == 1:
+            model = modeling.models.Polynomial1D(coefficients.size - 1)
+            model.parameters = coefficients
+        elif n_dim == 2:
+            shape = coefficients.shape
+            degree = shape[0] - 1
+            if shape[0] != shape[1]:
+                raise TypeError("Coefficients must be an (n+1, n+1) matrix")
+
+            coeffs = {}
+            for i in range(shape[0]):
+                for j in range(shape[0]):
+                    if i + j < degree + 1:
+                        name = 'c' + str(i) + '_' +str(j)
+                        coeffs[name] = coefficients[i, j]
+            model = modeling.models.Polynomial2D(degree, **coeffs)
```

Comment:
Yeah, I agree individual coeffs is fine.  Maybe supporting a coefficient matrix is nice too, but optional.  I just meant basically what you wrote in the last sentence, as well as a more flexible template for generating coefficient names (this goes back a while now to a discussion we had about SIP coefficients, for example).  This will be in astropy.modeling at some point...